src/util/usertools.c

```diff
@@ -835,3 +835,62 @@ int sss_output_fqname(TALLOC_CTX *mem_ctx,
     talloc_zfree(tmp_ctx);
     return ret;
 }
+
+static void sss_sssd_user_uid_and_gid(uid_t *uid, gid_t *gid)
+{
+    static uid_t sssd_uid;
+    static uid_t sssd_gid;
+    static bool resolved = false;
+
+    errno_t ret;
+
+    if (! resolved) {
+        ret = sss_user_by_name_or_uid(SSSD_USER, &sssd_uid, &sssd_gid);
+        if (ret != EOK) {
+            DEBUG(SSSDBG_CRIT_FAILURE, "failed to get sssd user (" SSSD_USER ") uid/gid, using root\n");
+            *uid = 0;
+            *gid = 0;
+            return;
+        } else {
+            resolved = true;
+        }
+    }
+
+    if (uid != NULL) {
+        *uid = sssd_uid;
+    }
+
+    if (gid != NULL) {
+        *gid = sssd_gid;
+    }
+}
+
+uid_t sss_sssd_user_uid(void)
+{
+    uid_t uid;
+    sss_sssd_user_uid_and_gid(&uid, NULL);
+    return uid;
+}
+
+gid_t sss_sssd_user_gid(void)
+{
+    gid_t gid;
+    sss_sssd_user_uid_and_gid(NULL, &gid);
+    return gid;
+}
+
+void sss_set_sssd_user_eid(void)
+{
+    if (geteuid() == 0) {
+        seteuid(sss_sssd_user_uid());
+        setegid(sss_sssd_user_gid());
+    }
+}
+
+void sss_restore_sssd_user_eid(void)
+{
+    if (getuid() == 0) {
+        seteuid(getuid());
+        setegid(getgid());
+    }
+}
```

Comment:
Let me see if I understood this part of the code correctly. If the `root` user owns the sssd process, then we change the effective id temporarily to `sssd`, connect to the ldb database, and then, revert the id to `root`. Is that correct?

Reply:
Yes, this way the "config.ldb" file is directly created with proper ownership in case the file is not present.